Simplify worker registration lifecycle.

jonathanknowles opened the pull request:
# Issue Number

#1292 

# Overview

This PR adjusts the worker registration life cycle so that creating and registering a worker is now a single operation, rather than a two-step operation.

Whereas previously callers would need to first call `newWorker` followed by `insert`, they now only need to call `register`.

In addition, we now provide automatic garbage collection of stale worker references from the registry, so that attempting to look up a previously-registered but prematurely-terminated thread will result in `Nothing`.

jonathanknowles committed:
Adjust code to meet coding standard.

jonathanknowles committed:
Add function `unregister` to module `Registry`.

This function allows a caller to unregister a worker without killing the
associated thread.

jonathanknowles committed:
Make creating and registering a worker into a single operation.

This change adjusts the worker registration lifecycle so that it's no
longer necessary to create and register a worker in two steps. Instead,
a single call to `register` will both create a worker and register it
within the specified registry.

In addition, we now provide automatic garbage collection of stale worker
references from the registry when worker threads terminate prematurely.

jonathanknowles committed:
Use function `create` to simplify function `Registry.register`.

KtorZ commented on lib/core/src/Cardano/Wallet/Registry.hs:
```diff
@@ -137,6 +136,16 @@ insert
+    -> key
+    -> IO ()
+unregister (WorkerRegistry mvar) k =
+    modifyMVar_ mvar (pure . Map.delete k)
```
Having both `unregister` and `remove` makes the module API somewhat non-orthogonal. What would you think of revising a bit the interface here by: 

- Keeping only one of the two, likely (`unregister` as it sounds more dual to `register`)
- Having `unregister` returns the actual `Worker key resource` when removed from the registry 
- Let the caller of `unregister` kill the worker thread itself

So, on call sites, instead of:

```hs
    liftHandler $ withWorkerCtx @_ @s @k ctx wid throwE $ \_ -> pure ()
    liftIO $ Registry.remove re wid
```

We would have:

```hs
    liftIO $ Registry.unregister re wid >>= \case
        Nothing  -> liftHandler $ throwE $ ErrNoSuchWallet wid
        Just wrk -> liftIO $ killThread $ workerThread wrk
```

which actually feels way less awkward that the artificial 'withWorkerCtx' call and make the whole thing much more explicit.

jonathanknowles replied:
@KtorZ I like the general idea, though I have one suggestion.

Since the registry is responsible for handling thread creation, perhaps it should also be responsible for thread destruction?

How about the following (names can be changed):
```hs
-- Exported:
unregister
    :: Ord key
    => WorkerRegistry key resource
    -> key
    -> IO ()  
unregister registry k =
    delete registry k >>= traverse_ (killThread . workerThread)

-- Internal operation, not exported:
delete
    :: Ord key
    => WorkerRegistry key resource
    -> key
    -> IO (Maybe (Worker key resource))   
delete (WorkerRegistry mvar) k = do   
    mWorker <- Map.lookup k <$> readMVar mvar 
    modifyMVar_ mvar (pure . Map.delete k)
    pure mWorker 
```

Then, `deleteWallet` can be:
```hs
deleteWallet ctx (ApiT wid) = 
   liftHandler $ withWorkerCtx @_ @s @k ctx wid throwE $ \_ -> do 
      liftIO $ Registry.unregister re wid  
      liftIO $ removeDatabase df wid  
      return NoContent   
```

jonathanknowles replied:
@KtorZ 

I've revised the interface of `Registry` so that `register` and `unregister` are **dual** operations of one another.

* `register` is responsible for:
    1. adding an entry into the registry map.
    2. thread creation.
* `unregister` is responsible for:
    1. removing an entry from the registry map.
    2. thread destruction.

See commit f0d5cbd612832e2757c7f8105d81a71362a958c9.

KtorZ commented on lib/core/src/Cardano/Wallet/Registry.hs:
```diff
@@ -190,44 +200,51 @@ defaultWorkerAfter tr = traceWith tr . \case
         )
     => ctx
     -> key
+    -> WorkerRegistry key resource
```
Side-note, all function working with the registry takes the registry as first argument. It'd be more consistent to also do this for this new function.

jonathanknowles replied:
@KtorZ wrote:
> Side-note, all function working with the registry takes the registry as first argument. It'd be more consistent to also do this for this new function.

Fixed in commit af7722992e91af4d8e8ccf5cc8826f8748cc6949.

KtorZ commented on lib/core/src/Cardano/Wallet/Registry.hs:
```diff
@@ -190,44 +200,51 @@ defaultWorkerAfter tr = traceWith tr . \case
+        (registry `unregister` k)
+        (finally
+            (tryPutMVar mvar Nothing)
+            (after tr e))
```
I do not understand the use of both `finally` here :thinking: ? These calls (`unregister` & `trryPutMVar`) are safe / don't throw, do there's no reason to wrap them in `finally` clauses, unless I am missing something?
The `cleanup` could be simply written:

```hs
cleanup mvar e = do
    registry `unregister` k
    tryPutMVar mvar Nothing
    after tr e
```

jonathanknowles replied:
@KtorZ wrote:
> I do not understand the use of both `finally` here thinking ? These calls (`unregister` & `trryPutMVar`) are safe / don't throw, do there's no reason to wrap them in `finally` clauses, unless I am missing something?

Thanks for spotting this. I think this was just a result of my paranoia while testing.

Fixed in commit 3bae4916ac3a26b756e70e7602d09a09f2178818.

jonathanknowles committed:
Change the parameter order for `Registry.register`.

Make the first parameter be a value of type `Registry`, for consistency
with other functions in the `Registry` module.

jonathanknowles committed:
Remove unnecessary use of `finally`.

The calls to `register` and `tryPutMVar` are safe, so it is unnecessary
to wrap them in `finally` here.

jonathanknowles committed:
Revise `Registry` so that `register` and `unregister` are dual operations.

The registry remains responsible for both thread creation (on
registration) and thread destruction (at the point of unregistration).

Anviking reviewed (approved):
Lgtm!

> In addition, we now provide automatic garbage collection of stale worker references from the registry, so that attempting to look up a previously-registered but prematurely-terminated thread will result in Nothing.

So if a worker terminates, it will be removed from the in-memory registry?

`listWallets` uses the in-memory Registry keys as source. If a worker *were* to terminate, the wallet would then go missing in Daedalus. Is that the most desirable behaviour? 🤔
